### ISSUE TITLE
Experimental diagnostics false positive on Box::pin(async move { ... })

### ISSUE THREAD

**Issue by Keavon** [rust-lang/rust-analyzer#19788]
**rust-analyzer version**:

```
rust-analyzer version: 0.3.2457-standalone (2bafe9d96c 2025-05-10) [c:\Users\Keavon\.vscode\extensions\rust-lang.rust-analyzer-0.3.2457-win32-x64\server\rust-analyzer.exe]
```

**rustc version**:
```
rustc 1.86.0 (05f9846f8 2025-03-31)
```

**editor or extension**:
VS Code, extension version `0.3.2457`

**relevant settings**:

![Image](https://github.com/user-attachments/assets/63c32aec-bcc0-49c6-83b6-e195274565c2)

**code snippet to reproduce**:

This minimal reproduction does, unfortunately, rely on one external crate: https://crates.io/crates/dyn-any But it is otherwise self-contained, you can copy this into a `lib.rs` file of a new project. Just add `dyn-any = "0.3.1"` to `[dependencies]` in `Cargo.toml` and put the following in `lib.rs`:

```rust
use core::future::Future;
use dyn_any::{DynAny, StaticTypeSized};
use std::pin::Pin;
use std::sync::Arc;

pub type Any<'n> = Box<dyn DynAny<'n> + 'n + Send>;
pub type DynFuture<'n, T> = Pin<Box<dyn Future<Output = T> + 'n + Send>>;
pub type FutureAny<'n> = DynFuture<'n, Any<'n>>;
pub type TypeErasedNode<'n> = dyn for<'i> NodeIO<'i, Any<'i>, Output = FutureAny<'i>> + 'n + Send + Sync;
pub type TypeErasedRef<'n> = &'n TypeErasedNode<'n>;
pub type SharedNodeContainer = Arc<TypeErasedRef<'static>>;

pub trait NodeIO<'i, Input>: Node<'i, Input>
where
	Self::Output: 'i + StaticTypeSized,
	Input: StaticTypeSized,
{
}

pub trait Node<'i, Input> {
	type Output: 'i;
	fn eval(&'i self, input: Input) -> Self::Output;
}

pub struct ComposeTypeErased {
	first: SharedNodeContainer,
	second: SharedNodeContainer,
}

impl<'i> Node<'i, Any<'i>> for ComposeTypeErased {
	type Output = DynFuture<'i, Any<'i>>;
	fn eval(&'i self, input: Any<'i>) -> Self::Output {
		// expected Pin<Box<dyn  Future<Output = Box<dyn DynAny<'_> + Send, Global>> + Send + 'i, Global>>,
		// found    Pin<Box<impl Future<Output = Box<dyn DynAny<'_> + Send, Global>>            , Global>>
		//                  ^^^^                                                    ^^^^^^^^^^^^
		Box::pin(async move {
			let arg = self.first.eval(input).await;
			self.second.eval(arg).await
		})
	}
}
```

![Image](https://github.com/user-attachments/assets/a96e17e7-2800-4080-b5ca-82b80a0a4a2c)

This passes `cargo check` fine:

```
C:\Users\Keavon\Desktop\reproduction>cargo check
    Checking reproduction v0.1.0 (C:\Users\Keavon\Desktop\reproduction)
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.11s
```

**Comment by ChayimFriedman2** [comment]
Yeah we know we have problems with our coercion setup, this will be fixed when we switch to the new trait solver. CC @ShoyuVanilla just in case we can do something now.

**Comment by ShoyuVanilla** [comment]
> Yeah we know we have problems with our coercion setup, this will be fixed when we switch to the new trait solver. CC [@ShoyuVanilla](https://github.com/ShoyuVanilla) just in case we can do something now.

I think that you are right that this is yet another next-solver-resolves issue but I'll look into this just in case

**Comment by ChayimFriedman2** [comment]
Probably fixed now.
